src/iop/hazeremoval.c

```diff
@@ -0,0 +1,973 @@
+  }
+  return first;
+}
+
+// quick select algorithm, arranges the range [first, last) such that
+// the element pointed to by nth is the same as the element that would
+// be in that position if the entire range [first, last) had been
+// sorted, additionally, none of the elements in the range [nth, last)
+// is less than any of the elements in the range [first, nth)
+void quick_select(float *first, float *nth, float *last) {
+  if (first == last)
+    return;
+  for ( ; ; ) {
+    float *p1 = first;
+    float *pivot = first + (last - first)/2;
+    float *p3 = last - 1;
+    if (!(*p1 < *pivot))
+      swap_pointer_f(p1, pivot);
+    if (!(*p1 < *p3))
+      swap_pointer_f(p1, p3);
+    if (!(*pivot < *p3))
+      swap_pointer_f(pivot, p3);
+    pointer_swap_f(pivot, last-1);  // move pivot to end
+    partition(first, last - 1, *(last - 1));
+    pointer_swap_f(last - 1, pivot);  // move pivot to its final place
+    if (nth == pivot)
+      break;
+    else if (nth < pivot)
+      last = pivot;
+    else
+      first = pivot + 1;
+  }
+}
+
+// calculate diffusive ambient light and the maximal depth in the image
+// depth is estimaged by the local amount of haze and given in units of the
+// characteristic haze depth, i.e., the distance over which object light is
+// reduced by the factor exp(-1)
+static float ambient_light(rgb_image img, int w1, rgb_pixel *pA0)
+{
+  const float dark_channel_quantil = 0.95f;  // quantil for determing the most hazy pixels
+  const float bright_quantil = 0.95f;  // quantil for determing the brightest pixels among the most hazy pixels
+  int width = img.width;
+  int height = img.height;
+  size_t size = (size_t)width * (size_t)height;
+  // calculate dark channel, which is an estimate for local amount of haze
+  gray_image dark_ch = new_gray_image(width, height);
+  dark_channel(img, dark_ch, w1);
+  // determine the brightest pixels among the most hazy pixels
+  gray_image bright_hazy = new_gray_image(width, height);
+  copy_gray_image(dark_ch, bright_hazy);
+  size_t p = size * dark_channel_quantil;
+  quick_select(bright_hazy.data, bright_hazy.data + p, bright_hazy.data + size);
+  float crit_haze_level = bright_hazy.data[p];
+  size_t N_most_hazy = 0;
+  for (size_t i = 0; i < size; i++)
+    if (dark_ch.data[i] >= crit_haze_level)
+    {
+      const float *pixel_in = get_rgb_pixel(img, i);
+      // next line prevents parallelization via OpenMP
+      bright_hazy.data[N_most_hazy] = pixel_in[0] + pixel_in[1] + pixel_in[2];
+      N_most_hazy++;
+    }
+  p = N_most_hazy * bright_quantil;
+  quick_select(bright_hazy.data, bright_hazy.data + p, bright_hazy.data + N_most_hazy);
+  float crit_brightness = bright_hazy.data[p];
+  free_gray_image(&bright_hazy);
+  // average over the brightest pixels among the most hazy pixels to
+  // estimate the diffusive ambient light
+  float A0[3] = {0, 0, 0};
+  size_t N_bright_hazy = 0;
+#ifdef _OPENMP
+#pragma omp parallel for reduction(+: N_bright_hazy, A0[:3] )
```

Comment:
Array reduction is a nice trick!
However it was added in `OpenMP 4.5`, which is only supported in `GCC 6.1+`
Currently we depend on `GCC 4.8`, i.e. only `OpenMP 3.1` can be unconditionally used.
http://www.openmp.org/resources/openmp-compilers/
https://godbolt.org/g/a6v33Y